metadata/src/main/java/org/apache/kafka/metadata/properties/MetaPropertiesEnsemble.java

```diff
@@ -0,0 +1,575 @@
+                if (!(uuid.toString().startsWith("-") || DirectoryId.reserved(uuid))) {
+                    boolean duplicate = false;
+                    for (MetaProperties metaProps : logDirProps.values()) {
+                        if (metaProps.directoryId().equals(Optional.of(uuid))) {
+                            duplicate = true;
+                            break;
+                        }
+                    }
+                    if (!duplicate) {
+                        return uuid;
+                    }
+                }
+            }
+        }
+
+        /**
+         * Verify that we have set up the Copier correctly.
+         *
+         * @throws RuntimeException if a directory appears in more than one category.
+         */
+        public void verify() {
+            for (String logDir : emptyLogDirs) {
+                if (errorLogDirs.contains(logDir)) {
+                    throw new RuntimeException("Error: log directory " + logDir +
+                        " is in both emptyLogDirs and errorLogDirs.");
+                }
+                if (logDirProps.containsKey(logDir)) {
+                    throw new RuntimeException("Error: log directory " + logDir +
+                        " is in both emptyLogDirs and logDirProps.");
+                }
+            }
+            for (String logDir : errorLogDirs) {
+                if (logDirProps.containsKey(logDir)) {
+                    throw new RuntimeException("Error: log directory " + logDir +
+                            " is in both errorLogDirs and logDirProps.");
+                }
+            }
+            metaLogDir().ifPresent(m -> {
+                if (!(emptyLogDirs.contains(m) ||
+                        logDirProps.containsKey(m) ||
+                        errorLogDirs.contains(m))) {
+                    throw new RuntimeException("Error: metaLogDir " + m + " does not appear " +
+                        "in emptyLogDirs, errorLogDirs, or logDirProps.");
+                }
+            });
+        }
+
+        /**
+         * Write any changed log directories out to disk.
+         */
+        public void writeLogDirChanges() {
+            writeLogDirChanges((metaPropsPath, e) -> {
+                LOG.error("Error while writing meta.properties file {}", metaPropsPath, e);
+            });
+        }
+
+        /**
+         * Write any changed log directories out to disk.
+         *
+         * @param errorHandler  A handler that will be called with (logDir, exception) any time we
+         *                      hit an exception writing out a meta.properties file.
+         */
+        public void writeLogDirChanges(BiConsumer<String, IOException> errorHandler) {
+            Map<String, MetaProperties> newOrChanged = new HashMap<>();
+            HashSet<String> newSet = new HashSet<>();
+            for (Entry<String, MetaProperties> entry : prev.logDirProps().entrySet()) {
```

Comment:
This is getting the entrySet of an Ensemble which doesn't include directoryId() so adding a directoryId doesn't show up as changed and thus it won't get committed to disk.

Reply:
Well, there are two entrySets here: the copier's, and the prev ensemble's. The MetaProperties object from one won't match the other, so it will get rewritten.